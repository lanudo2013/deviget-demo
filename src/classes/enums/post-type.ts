export enum PostType {
    IMAGE, VIDEO, LINK, SELF
}

export function fromPostType(val: string) : PostType | null {
    if (val) {
        val = val.toLowerCase();
        switch(val) {
            case 'image':
                return PostType.IMAGE;
            case 'hosted:video':
                return PostType.VIDEO;
            case 'self':
                return PostType.SELF;
            case 'link':
                return PostType.LINK;
            default:
                return null;
        }
    }
    return PostType.IMAGE;
}