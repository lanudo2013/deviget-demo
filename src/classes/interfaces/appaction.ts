import { Action } from "redux";

export interface AppAction<T> extends Action {
    payload: T;
}