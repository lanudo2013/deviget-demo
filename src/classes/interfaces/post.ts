export interface Post {
    title: string;
    thumbnailUrl: string;
    createdTime: Date;
    author: string;
    id: string;
    fullPictureUrl: string;
    subreddit: string;
    thumbnailDims?: {width: number, height: number};
    numberOfComments: number;
}