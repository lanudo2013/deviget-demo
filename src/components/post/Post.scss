.PostContainer {
    display: flex;
    flex-direction: column;
    margin-left: 10px;
    border-radius: 20px;
    margin-right: 10px;
    margin-top: 3px;
    padding: 20px;
    margin-bottom: 10px;
    box-shadow: 2px 1px 3px 2px rgb(191 191 191 / 42%);
    animation-duration: 0.8s !important;

    .Header {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        .Author {
            font-size: 20px;
            font-weight: bold;
            line-height: 10px;
        }
    }
    .Body {
        min-height: 90px;
        display: flex;
        margin-top: 5px;
        align-items: center;
        border: solid 0px black;
        .Thumbnail {
        }
        .Title {
            vertical-align: center;
            margin-left: 10px;
            align-self: center;
        }
    }
    .Footer {
        margin-top: 7px;
        display: flex;
        justify-content: space-between;
        .DismissButton {
            padding-top: 1px;
            padding-bottom: 1px;
        }
        .CommentsNum {
            color: rgb(165, 165, 41);
            font-size: 16px;
            font-weight: bold;
        }
    }
}