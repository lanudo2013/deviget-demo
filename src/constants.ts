export class Constants {
    public static readonly REDDIT_BASE_URL = `https://www.reddit.com/top.json`;
    public static readonly HTTP_TIMEOUT = 40000;
    public static ERRORS = {
        TIMEOUT_CODE: -1
    }

    public static REDUX_ACTIONS = {
        FETCH_REQUEST_POSTS: 'fetch-request-posts',
        SUCCESS_REQUEST_POSTS: 'success-request-posts',
        FAIL_REQUEST_POSTS: 'fail-request-posts',

        DISMISS_POST: 'dismiss-post',
        SELECT_POST: 'select-post',

        READ_POSTS_IDS: 'read-posts-ids',
        DISMISSED_POSTS_IDS: 'dismissed-posts-ids',
        UPDATE_CURRENT_ERROR: 'update-current-error',

        UPDATE_DONE_DISMISS_DATA: 'update-done-dismiss-data'
    }

    public static POST_LIST_PAGE_SIZES = [25, 50, 100];
    public static MOBILE_WIDHT_LIMIT_PX = 1200;

    public static APP_MESSAGES = {
        HTTP_TIMEOUT: 'Connection timeout',
        ERROR_GET_POSTS: 'Unable to get posts',
        REFRESH_BUTTON: 'Refresh',
        DISMISS_ALL_BUTTON: 'Dismiss All',
        ERROR_TITLE: 'Error',
        DISMISS_BUTTON: 'Dismiss Post',
        LOADING: 'Loading',
        ERROR_DISMISS_POST: 'Unable to dismiss post',
        ERROR_DISMISS_POSTS: 'Unable to dismiss posts',
        ERROR_SAVE_READ_POST: "Unable to save read post state",
        PAGE_SIZE_PLC: "Page Size"

    }
}