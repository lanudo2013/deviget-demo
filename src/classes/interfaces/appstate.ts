import { DismissData } from "./dismiss-data";
import { Post } from "./post";

export interface AppState {
    posts: Post[];
    selectedPost: Post | undefined | null;
    fetchingPosts: boolean;
    currentError: string;
    dismissData: DismissData | undefined;
    postsRead: string[];
}