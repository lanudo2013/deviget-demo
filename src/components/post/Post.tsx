import { Post } from "../../classes/interfaces/post";
import React, { useState } from 'react';
import moment from  'moment';
import './Post.scss';
import { Button } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/CancelOutlined';

const DAY_TIME = 1000 * 60 * 60 * 24;
const HOUR_TIME = 1000 * 60 * 60;
const MINUTE_TIME = 1000 * 60;

interface PostProps {
    post: Post;
    currentDate: Date;
    onPress: (p: Post) => void;
    onDismiss: (post: Post) => void;
}
export const PostUI = function(props: PostProps) {
    const [pressAnimation, setPressAnimation] = useState<boolean>(false);
    const [dismissAnimation, setDismissAnimation] = useState<boolean>(false);
    const formatCreated = React.useCallback((dt: Date) => {
        const now = props.currentDate;
        const diff = now.getTime() - dt.getTime();
        if (diff < HOUR_TIME) {
            return `${Math.floor(diff / MINUTE_TIME)} minutes ago`;
        } else if (diff < DAY_TIME) {
            return `${Math.floor(diff / HOUR_TIME)} hours ago`;
        } else if (diff < 5 * DAY_TIME) {
            return `${Math.floor(diff / DAY_TIME)} days, ${(diff % DAY_TIME) / HOUR_TIME} hours ago`;
        } else {
            return `${Math.floor(diff / DAY_TIME)} days ago`;
        }
    }, []);

    const dismissButtonStyle = React.useMemo(() => ({paddingTop: '1px', paddingBottom: '1px', fontSize: '12px'}), []);

    const p = props.post;
    const pressPost = React.useCallback((e) => {
        setPressAnimation(true);
        props.onPress && props.onPress(p);
    }, [props.onPress, p]);

    
    const animEnds = React.useCallback((e) => {
        setPressAnimation(false);
        if (dismissAnimation) {
            props.onDismiss && props.onDismiss(p);
        }
    }, [p, props.onDismiss]);

    const pressDismiss = React.useCallback((e: any) => {
        e.stopPropagation();
        setDismissAnimation(true);
    }, []);

    if (!props.post) {
        return null;
    }
    


    return <div className={'PostContainer ' + (pressAnimation ? ' animate__animated animate__headShake' : '') + (dismissAnimation ? 'animate__animated animate__fadeOut' : '')} onClick={pressPost} onAnimationEnd={animEnds}>
        <div className="Header">
            <span className="Author">{p.author}</span>
            <span className="CreatedAt">{formatCreated(p.createdTime)}</span>
        </div>
        <div className="Body">
            {p.thumbnailUrl ? <img src={p.thumbnailUrl} width={100} height={100} className="Thumbnail" /> : null}
            <span className="Title">{p.title}</span>
        </div>
        <div className="Footer">
            <Button onClick={pressDismiss} color="default" size={'small'} style={dismissButtonStyle} variant="contained" startIcon={<DeleteIcon />}>Dismiss post</Button>
            <span className="CommentsNum">{p.numberOfComments} comment/s</span>
        </div>
    </div>;
};