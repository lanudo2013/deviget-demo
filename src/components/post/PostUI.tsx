import { Post } from '../../classes/interfaces/post';
import React, { ForwardedRef, RefObject, useEffect, useImperativeHandle, useRef, useState } from 'react';
import moment from 'moment';
import './Post.scss';
import { Button } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/CancelOutlined';
import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline';
import SaveIcon from '@material-ui/icons/SaveOutlined';
import Icon from '@material-ui/core/Icon';
import { Constants } from '../../constants';
import Intl from 'intl';
import { AnyIfEmpty } from 'react-redux';
import { PostType } from '../../classes/enums/post-type';

const DAY_TIME = 1000 * 60 * 60 * 24;
const HOUR_TIME = 1000 * 60 * 60;
const MINUTE_TIME = 1000 * 60;

interface PostProps {
    post: Post;
    currentDate: Date;
    onPress: (p: Post) => void;
    onPressSave?: (p: Post) => void;
    onPressRemoved?: (p: Post) => void;
    read?: boolean;
    canRemove?: boolean;
    saved?: boolean;
    onPressDismiss: (post: Post) => void;
}
export interface PostRef {
    fadeOut: () => Promise<any>;
    id: string;
    slideOut: (delay?: number) => Promise<any>;
}
const PostUIFn = function (props: PostProps, ref1: ForwardedRef<any>) {
    const [pressAnimation, setPressAnimation] = useState<boolean>(false);
    const [fadeOutAnimation, setDismissFadeOutAnimation] = useState<boolean>(false);
    const [slideOutAnimation, setDismissSlideOutAnimation] = useState<boolean>(false);
    const [hidden, setHidden] = useState<boolean>(false);
    const resolveFadeOutAnim = useRef<any>(null);
    const resolveSlideAnim = useRef<any>(null);
    const formatCreatedAt = React.useCallback(
        (dt: Date) => {
            const now = props.currentDate;
            if (now && dt) {
                const diff = now.getTime() + now.getTimezoneOffset() * 1000 * 60 - dt.getTime();
                const days = Math.floor(diff / DAY_TIME);
                const minutes = Math.floor(diff / MINUTE_TIME);
                const hours = Math.floor(diff / HOUR_TIME);
                if (diff > 0) {
                    if (diff < MINUTE_TIME) {
                        return 'Seconds ago';
                    } else if (diff < HOUR_TIME) {
                        return `${minutes} minute${minutes > 1 ? 's' : ''} ago`;
                    } else if (diff < DAY_TIME) {
                        return `${hours} hour${hours > 1 ? 's' : ''} ago`;
                    } else if (diff < 5 * DAY_TIME) {
                        const hoursDiff = Math.floor((diff % DAY_TIME) / HOUR_TIME);
                        if (hoursDiff > 0) {
                            return `${days} day${days > 1 ? 's' : ''}, ${hoursDiff} hour${hoursDiff > 1 ? 's' : ''} ago`;
                        } else {
                            return `${days} day${days > 1 ? 's' : ''} ago`;
                        }
                    } else {
                        return `${days} day${days > 1 ? 's' : ''} ago`;
                    }
                }
            }
            return '';
        },
        [props.currentDate]
    );

    const p = props.post;
    const dismissButtonStyle = React.useMemo(
        () => ({ paddingTop: '1px', paddingBottom: '1px', fontSize: '12px', marginRight: '5px' }),
        []
    );
    useImperativeHandle(
        ref1,
        () => ({
            fadeOut: () => {
                if (resolveFadeOutAnim.current) {
                    return Promise.reject('Animating');
                }

                setDismissFadeOutAnimation(true);
                return new Promise((res) => {
                    resolveFadeOutAnim.current = res;
                });
            },
            slideOut: (delay?: number) => {
                if (resolveSlideAnim.current) {
                    return Promise.reject('Animating');
                }
                if (!delay) {
                    setDismissSlideOutAnimation(true);
                }
                return new Promise((res) => {
                    resolveSlideAnim.current = res;
                    if (delay) {
                        setTimeout(() => {
                            setDismissSlideOutAnimation(true);
                        }, delay);
                    }
                });
            },
            id: p.id
        }),
        [p]
    );

    const formatter = React.useMemo(() => {
        return new Intl.NumberFormat('en-IR');
    }, []);

    const pressPost = React.useCallback(
        (e) => {
            setPressAnimation(true);
            props.onPress && props.onPress(p);
        },
        [props.onPress, p]
    );

    const pressSave = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressSave && props.onPressSave(p);
        },
        [props.onPressSave, p]
    );

    const animEnds = React.useCallback(
        (e) => {
            setPressAnimation(false);
            if (fadeOutAnimation && resolveFadeOutAnim.current) {
                resolveFadeOutAnim.current(true);
            } else if (slideOutAnimation && resolveSlideAnim.current) {
                setHidden(true);
                resolveSlideAnim.current(true);
            }
            resolveSlideAnim.current = null;
            resolveFadeOutAnim.current = null;
            setDismissSlideOutAnimation(false);
            setDismissFadeOutAnimation(false);
        },
        [p, fadeOutAnimation, slideOutAnimation]
    );

    const pressDismiss = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressDismiss && props.onPressDismiss(p);
        },
        [props.onPressDismiss, p]
    );

    const pressRemoved = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressRemoved && props.onPressRemoved(p);
        },
        [props.onPressRemoved, p]
    );

    const animationClassStr = React.useMemo(() => {
        if (pressAnimation) {
            return ' animate__animated animate__headShake';
        } else if (fadeOutAnimation) {
            return ' animate__animated animate__fadeOut';
        } else if (slideOutAnimation) {
            return ' animate__animated animate__slideOutLeft';
        } else {
            return '';
        }
    }, [slideOutAnimation, fadeOutAnimation, pressAnimation]);

    const openFullSizeImage = React.useCallback(
        (e) => {
            const val = p.postUrl || '';
            if (p.postType === PostType.IMAGE && (val.endsWith('.jpg') || val.endsWith('.png') || val.endsWith('.gif') || val.endsWith('.jpeg'))) {
                e.stopPropagation();
                window.open(p.postUrl, '_blank');
            }
        },
        [p]
    );

    if (!props.post) {
        return null;
    }

    return (
        <div
            className={'PostContainer ' + animationClassStr}
            style={hidden ? { visibility: 'hidden' } : {}}
            onClick={pressPost}
            onAnimationEnd={animEnds}
        >
            <div className="Header">
                <div className="Header-left">
                    {!props.read ? <Icon className="ReadIcon">fiber_manual_record</Icon> : null}
                    <span className="Author" style={props.read ? { left: 0 } : null}>
                        {p.author}
                    </span>
                </div>
                <div className="Header-right">
                    {props.saved ? <Icon className="SavedIcon">save</Icon> : null}
                    <span className="CreatedAt">{formatCreatedAt(p.createdTime)}</span>
                </div>
            </div>
            <div className="Body">
                {p.thumbnailUrl ? (
                    <img
                        src={p.thumbnailUrl}
                        width={100}
                        height={100}
                        onClick={openFullSizeImage}
                        className={'Thumbnail ' + (p.postType === PostType.IMAGE ? 'Clickable' : '')}
                    />
                ) : null}
                <span className="Title">{p.title}</span>
            </div>
            <div className="Footer">
                <div className="PostOptions">
                    {!props.saved ? (
                        <>
                            <Button
                                onClick={pressDismiss}
                                color="default"
                                size={'small'}
                                style={dismissButtonStyle}
                                variant="contained"
                                className="DismissButton"
                                startIcon={<DeleteIcon />}
                            >
                                {Constants.APP_MESSAGES.DISMISS_BUTTON}
                            </Button>
                            <Button
                                onClick={pressSave}
                                color="default"
                                size={'small'}
                                style={dismissButtonStyle}
                                variant="contained"
                                className="SaveButton"
                                startIcon={<SaveIcon />}
                            >
                                {Constants.APP_MESSAGES.SAVE_BUTTON}
                            </Button>
                        </>
                    ) : (
                        <>
                            {props.canRemove ? (
                                <Button
                                    onClick={pressRemoved}
                                    color="default"
                                    size={'small'}
                                    style={dismissButtonStyle}
                                    variant="contained"
                                    className="RemoveButton"
                                    startIcon={<DeleteOutlineIcon />}
                                >
                                    {Constants.APP_MESSAGES.REMOVE_BUTTON}
                                </Button>
                            ) : (
                                <Button
                                    onClick={pressDismiss}
                                    color="default"
                                    size={'small'}
                                    style={dismissButtonStyle}
                                    variant="contained"
                                    className="DismissButton"
                                    startIcon={<DeleteIcon />}
                                >
                                    {Constants.APP_MESSAGES.DISMISS_BUTTON}
                                </Button>
                            )}
                        </>
                    )}
                </div>

                <span className="CommentsNum">
                    {p.numberOfComments > 0 ? formatter.format(p.numberOfComments) + ' comment/s' : 'No comments'}
                </span>
            </div>
        </div>
    );
};
export const PostUI = React.forwardRef(PostUIFn);
