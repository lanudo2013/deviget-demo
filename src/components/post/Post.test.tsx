import { fromPostType, PostType } from '../../classes/enums/post-type';
import { Post } from '../../classes/interfaces/post';
import { mount, ReactWrapper } from "enzyme";
import moment from 'moment';
import React, { useRef } from 'react';
import renderer from 'react-test-renderer';
import { PostUI } from './PostUI';
import { Button } from '@material-ui/core';
import { Constants } from '../../constants';

describe('Post component', () => {
    let root: ReactWrapper;
    let listP: Post[];
    let onPressFn: (p: Post) => void;
    let onPressDismissFn: (p: Post) => void;
    let currentDate = new Date();

    function doBasicChecks(node: any) {
        const tree = renderer.create(node).toJSON();
        expect(tree).toMatchSnapshot();

        let domElement: ReactWrapper = root.find('div.PostContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostContainer > div.Header');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostContainer > div.Body');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostContainer > div.Footer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();

    }

    function checkAuthorTitleComments(p: Post, expNumCommentsText: string) {
        let domElement: ReactWrapper;
        domElement = root.find('span.Author');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual(p.author);

        domElement = root.find('span.Title');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual(p.title);

        domElement = root.find('span.CommentsNum');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual(expNumCommentsText);
    }

    function createPosts(): Post[] {
        return [{
            author: 'author1',
            createdTime: moment(currentDate).add(-4, 'days').add(-6, 'hours').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 1230,
            id: 'p1',
            postType: PostType.IMAGE,
            thumbnailUrl: '/p1-thumb.jpg',
            title: 'My new post 1',
            subreddit: 'Sub',
            postUrl: '/p1.jpg',
            postHtml: null
        },
        {
            author: 'author1',
            createdTime: moment(currentDate).add(-10, 'days').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 878,
            id: 'p2',
            postType: PostType.IMAGE,
            thumbnailUrl: '/p2-thumb.jpg',
            title: 'My new post 2',
            subreddit: 'Sub2',
            postUrl: '/p2.jpg',
            postHtml: null
        },
        {
            author: 'author2',
            createdTime: moment(currentDate).add(-2, 'minutes').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 900,
            id: 'p3',
            postType: PostType.LINK,
            thumbnailUrl: undefined,
            title: 'My new post 3',
            subreddit: 'Sub3',
            postUrl: '/p3.jpg',
            postHtml: null
        },
        {
            author: 'author3',
            createdTime: moment(currentDate).add(-12, 'seconds').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 0,
            id: 'p4',
            postType: PostType.SELF,
            thumbnailUrl: undefined,
            title: 'My new post 4',
            subreddit: 'Sub4',
            postUrl: undefined,
            postHtml: null
        }];
    }

    beforeEach((dn) => {
        listP = createPosts();
        onPressFn = jest.fn();
        onPressDismissFn = jest.fn();
        dn();
    });

    test('Select post type', () => {
        let val = fromPostType({post_hint: 'image'});
        expect(val).toEqual(PostType.IMAGE);
        val = fromPostType({is_video: true});
        expect(val).toEqual(PostType.VIDEO);
        val = fromPostType({});
        expect(val).toEqual(PostType.IMAGE);
        val = fromPostType({selftext_html: '<div></div>'});
        expect(val).toEqual(PostType.SELF);
    });
 
    test('Load normal post', () => 
    {
        const node = <PostUI post={listP[0]} currentDate={currentDate} onPress={onPressFn} onPressDismiss={onPressDismissFn}  read={false}/>;
        root = mount(node);
        doBasicChecks(node);
        checkAuthorTitleComments(listP[0], '1,230 comment/s');

        let domElement = root.find('span.CreatedAt');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual('4 days, 6 hours ago');

        domElement = root.find('div.Body > img.Thumbnail');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.get(0).props.src).toEqual(listP[0].thumbnailUrl);
    });

    test('Load normal post 2', () => 
    {
        const p = listP[1];
        const node = <PostUI post={p} currentDate={currentDate} onPress={onPressFn} onPressDismiss={onPressDismissFn}  read={false}/>;
        root = mount(node);
        doBasicChecks(node);
        checkAuthorTitleComments(p, '878 comment/s');

        let domElement = root.find('span.CreatedAt');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual('10 days ago');

        domElement = root.find('div.Body > img.Thumbnail');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.get(0).props.src).toEqual(p.thumbnailUrl);

        let buttonWrapper: ReactWrapper = root.find(Button);
        buttonWrapper.find('button').simulate('click');
    });

    test('Load post without picture', () => 
    {
        const p = listP[2];
        const node = <PostUI post={p} currentDate={currentDate} onPress={onPressFn} onPressDismiss={onPressDismissFn}  read={false}/>;
        root = mount(node);
        doBasicChecks(node);
        checkAuthorTitleComments(p, '900 comment/s');

        let domElement = root.find('span.CreatedAt');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual('2 minutes ago');

        domElement = root.find('div.Body > img.Thumbnail');
        expect(domElement.get(0)).toBeUndefined();
    });

    test('Press callbacks', () => 
    {
        const p = listP[3];
        const postRef = {current: {fadeOut: null, slideOut: null}} as any;
        const node = <PostUI post={p} currentDate={currentDate} onPress={onPressFn} onPressDismiss={onPressDismissFn} ref={postRef} read={false}/>;
        
        root = mount(node);
        doBasicChecks(node);
        checkAuthorTitleComments(p, 'No comments');

        let domElement = root.find('span.CreatedAt');
        expect(domElement.length).toEqual(1);
        expect(domElement.text()).toEqual('Seconds ago');

        let buttonWrapper: ReactWrapper = root.find(Button);
        buttonWrapper.find('button').simulate('click');

        expect(onPressDismissFn).toHaveBeenCalled();
        expect(postRef.current.fadeOut).toBeTruthy();
        expect(postRef.current.slideOut).toBeTruthy();

        buttonWrapper = root.find('div.PostContainer');
        buttonWrapper.simulate('click');

        expect(onPressFn).toHaveBeenCalled();
    });

    test('not read post to read post', () => {
        const p = listP[3];
        const postRef = {current: {fadeOut: null, slideOut: null}} as any;
        const node = <PostUI post={p} currentDate={currentDate} onPress={onPressFn} onPressDismiss={onPressDismissFn} ref={postRef} read={false}/>;
        
        root = mount(node);
        let domElement = root.find('div.Header > div.Header-left > .ReadIcon');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        root.setProps({read: true});
        domElement = root.find('div.Header > div.Header-left > .ReadIcon');
        expect(domElement.get(0)).toBeUndefined();

    })
});