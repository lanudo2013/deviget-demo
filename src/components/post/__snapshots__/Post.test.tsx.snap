// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component Load normal post 1`] = `
<div
  className="PostContainer "
  onAnimationEnd={[Function]}
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="Header"
  >
    <div
      className="Header-left"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root ReadIcon"
      >
        fiber_manual_record
      </span>
      <span
        className="Author"
        style={null}
      >
        author1
      </span>
    </div>
    <span
      className="CreatedAt"
    >
      4 days, 6 hours ago
    </span>
  </div>
  <div
    className="Body"
  >
    <img
      className="Thumbnail"
      height={100}
      src="/p1-thumb.jpg"
      width={100}
    />
    <span
      className="Title"
    >
      My new post 1
    </span>
  </div>
  <div
    className="Footer"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "12px",
          "paddingBottom": "1px",
          "paddingTop": "1px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
            />
          </svg>
        </span>
        Dismiss Post
      </span>
    </button>
    <span
      className="CommentsNum"
    >
      1,230 comment/s
    </span>
  </div>
</div>
`;

exports[`Post component Load normal post 2 1`] = `
<div
  className="PostContainer "
  onAnimationEnd={[Function]}
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="Header"
  >
    <div
      className="Header-left"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root ReadIcon"
      >
        fiber_manual_record
      </span>
      <span
        className="Author"
        style={null}
      >
        author1
      </span>
    </div>
    <span
      className="CreatedAt"
    >
      10 days ago
    </span>
  </div>
  <div
    className="Body"
  >
    <img
      className="Thumbnail"
      height={100}
      src="/p2-thumb.jpg"
      width={100}
    />
    <span
      className="Title"
    >
      My new post 2
    </span>
  </div>
  <div
    className="Footer"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "12px",
          "paddingBottom": "1px",
          "paddingTop": "1px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
            />
          </svg>
        </span>
        Dismiss Post
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <span
      className="CommentsNum"
    >
      878 comment/s
    </span>
  </div>
</div>
`;

exports[`Post component Load post without picture 1`] = `
<div
  className="PostContainer "
  onAnimationEnd={[Function]}
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="Header"
  >
    <div
      className="Header-left"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root ReadIcon"
      >
        fiber_manual_record
      </span>
      <span
        className="Author"
        style={null}
      >
        author2
      </span>
    </div>
    <span
      className="CreatedAt"
    >
      2 minutes ago
    </span>
  </div>
  <div
    className="Body"
  >
    <span
      className="Title"
    >
      My new post 3
    </span>
  </div>
  <div
    className="Footer"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "12px",
          "paddingBottom": "1px",
          "paddingTop": "1px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
            />
          </svg>
        </span>
        Dismiss Post
      </span>
    </button>
    <span
      className="CommentsNum"
    >
      900 comment/s
    </span>
  </div>
</div>
`;

exports[`Post component Press callbacks 1`] = `
<div
  className="PostContainer "
  onAnimationEnd={[Function]}
  onClick={[Function]}
  style={Object {}}
>
  <div
    className="Header"
  >
    <div
      className="Header-left"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root ReadIcon"
      >
        fiber_manual_record
      </span>
      <span
        className="Author"
        style={null}
      >
        author3
      </span>
    </div>
    <span
      className="CreatedAt"
    >
      Seconds ago
    </span>
  </div>
  <div
    className="Body"
  >
    <span
      className="Title"
    >
      My new post 4
    </span>
  </div>
  <div
    className="Footer"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "12px",
          "paddingBottom": "1px",
          "paddingTop": "1px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
            />
          </svg>
        </span>
        Dismiss Post
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <span
      className="CommentsNum"
    >
      No comments
    </span>
  </div>
</div>
`;
